### PR TITLE
Adding control properties API

### LINKED ISSUE #47: Dynamic Properties
It would be nice to have dynamic properties that become visible or hidden based on another property.

For instance, you can have a type property that drives subsequent properties that are visible, such as an additional label only present if that type were selected.

### PR DESCRIPTION
**Pull request by erickzanardo**
This PR adds the control properties API.

Fixes #47 

https://user-images.githubusercontent.com/835641/115445969-eb7b1d00-a1ec-11eb-8f32-2ef8bdf04b71.mov

### CODE REVIEW

**Review by tyemy — APPROVED** [review]

File: example/lib/widgets/message_card.dart [hunk 1]
```diff
@@ -0,0 +1,32 @@
+import 'package:flutter/material.dart';
+
+enum MessageCardType {
+  info,
+  error,
+}
+
+/// Naive widget to be an example of a little bit more
+/// complex widget that can contain different modes
+class MessageCard extends StatelessWidget {
```
Comment (jeffs-sf) [inline example/lib/widgets/message_card.dart:10]: **suggestion:**
```suggestion
/// Naive widget to be an example of a little bit more
/// complex one with different types.
class MessageCard extends StatelessWidget {
```

File: lib/src/widgets/properties_container.dart [hunk 2]
```diff
@@ -23,64 +23,92 @@ class PropertiesContainer extends StatefulWidget {
 class _PropertiesContainerState extends State<PropertiesContainer> {
   @override
   Widget build(BuildContext context) {
+    final List<Widget> children = [];
+
+    for (var entry in widget.currentChapter.ctx.properties.entries) {
```
Comment (jeffs-sf) [inline lib/src/widgets/properties_container.dart:28]: **suggestion:** we could break this out into a method

File: lib/src/story.dart [hunk 3]
```diff
@@ -19,14 +28,16 @@ class Property<T> {
...
+      {ControlProperty? controlProperty})
+      : super(name, defaultValue, controlProperty: controlProperty);
 }
 
 class DashbookContext {
   Map<String, Property> properties = {};
 
-  String textProperty(String name, String defaultValue) {
+  String textProperty(String name, String defaultValue,
+      {ControlProperty? controlProperty}) {
```
Comment (jeffs-sf) [inline lib/src/story.dart:40]: **chore (blocking):** we need to utilize the `controlProperty` here and pass it to the `Property` like we added for the other property types.
https://github.com/erickzanardo/dashbook/blob/0de95e01ae6898d27deb4e5f0b7fedcae763fad1/lib/src/story.dart#L33-L34

```dart
      final property = Property<String>(name, defaultValue,
          controlProperty: controlProperty);
      properties[name] = property;
```

**Review by jeffs-sf — CHANGES_REQUESTED** [review]
As the author of #47 this does exactly what I was looking for, thank you! There's some room for expansion later (maybe hiding if it doesn't match the type verses including if it does), but this helps tremendously.

There's one missing property that doesn't currently respect the `ControlProperty` but allows the consumer to define it (`TextProperty`) and some minor grammar suggestions. I added support for it locally to verify that passing in that `ControlProperty` to the `Property` works.

**Comment by erickzanardo** [comment]
Thanks for the review @tyemy and @jeffs-sf I have applied the suggestions.

**Pull request merged**